Adding Gulfstream conceptual overview (#568)
chainguard-dev/edu: content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/gulfstream-overview/index.md

![Diagram representing a single cluster running the following extensions: cert-manager, Istio, Tekton, Knative, Kafka Operator, Kpack, Flux, Prometheus Operator.](single_cluster.png)

CRDs are only made up of data — which is often schematized — but they are brought to life by controller processes and webhooks that conventionally run on the cluster itself, serving as a sort of extension to the built-in control plane processes. 

With traditional controller architectures, every cluster in a fleet runs every one of the clusters needed to run certain abstractions. This means that it's increasingly common for platform operations teams to be managing multiple extensions across many clusters, as in this diagram.

![Diagram similar to the first in this guide, showing multiple clusters across three area zones (us-west2, us-central3, and us-east1). However, this diagram highlights the extensions running on each cluster.](clusters_per_AZ_extensions.png)

As the number of clusters grows, so does the difficulty and cost of offering managed extensions to each of them, potentially growing into a complexity nightmare.

## Understanding Controllers, Reconcilers, and Control Planes

> With the spread of CRDs, it can be difficult to know how trustworthy a given extension is. When you extend Kubernetes with extensions based on CRDs, you install additional control planes which implement the semantics of their extension. A *control plane* consists of the various APIs that control the behavior of systems; in Kubernetes, the API server and its controllers form the control plane. 

Reviewer: This is a tiny nit, but what about subbing "extend" for something else since "extensions" is so nearby? I know they are technical terms so maybe disregard. One possibility I thought of:

"When you broaden the scope of Kubernetes with extensions..."

Author: hadn't even noticed that, great suggestion!

Knative's controller architecture separates the logic of controllers and reconcilers, as the following diagram represents.

![Diagram outlining a typical controller architecture. Within the controller are informers (each with a lister cache), a workqueue, clients, and a reconciler. The informers communicate with the workqueue, which passes events to the reconciler. The reconciler then communicates with the clients and lister cache as it performs reconciliation. Meanwhile, the informers and clients can communicate with an external API server.](controller_architecture.png)

In this architecture, the controller is aware of informers and the workqueue, sets up the various relationships between event triggers and the workqueue, and hands off clients and listers to the reconciler. Reconcilers know nothing about informers, they just know about the clients and listers in order to access what they need to enable reconciliation.

## Trusting trusted control planes

> [SLSA](https://slsa.dev/) makes several references to the idea of *trusted control planes*. Put simply, a trusted control plane is one that you trust in some way that you do not trust other control planes or workloads.

Reviewer: As the next paragraph clarifies, there are reasons for there to be trust in "trusted control planes." Could we maybe combine the paragraphs and remove the "in some way" phrasing to make it clearer?

Author: I'm having a really hard time losing the "in some way" phrasing, so what I did was add an example borrowed from MM's blog post on trusted control planes. This will tie it back to SLSA a bit more and provide more context for the following "supervisor processes" paragraph. I think it's a lot clearer than what I originally had

> You can think of trusted control planes as being akin to supervisor processes: they are often adjacent to normal workloads and can support them, for example by fetching external dependencies. Trusted control planes can also audit workloads, such as by signing provenance attestations, or authorize them to perform certain functions, as in the case of policy engines. 

Reviewer: My reading is that "authorize" should be parallel with "signing" here, so it should be the gerund

```suggestion
You can think of trusted control planes as being akin to supervisor processes: they are often adjacent to normal workloads and can support them, for example by fetching external dependencies. Trusted control planes can also audit workloads, such as by signing provenance attestations, or authorizing them to perform certain functions, as in the case of policy engines. 
```

Trusted control planes are a critical part of securing the software supply chain, whether it is the trusted control plane orchestrating aspects of your CI/CD pipeline or the one making policy decisions about what is allowed to run in your production cluster. 

In the context of Kubernetes, running trusted control planes presents a significant complication when it comes to the level of isolation Kubernetes provides between workloads. The weaker this isolation is, the less confidence you can have that a control plane workload has not been compromised. Strong isolation is sometimes called “hard multi-tenancy”, where the typical isolation in Kubernetes is often called “soft multi-tenancy”.

> One strategy used to isolate workloads is to run them in separate node pools. This way, even with a container escape, the workloads are running on different nodes. However, if the attacker gains access toAPI server credentials, they can potentially move laterally and try to compromise workloads on different nodes.

Reviewer: Space

```suggestion
One strategy used to isolate workloads is to run them in separate node pools. This way, even with a container escape, the workloads are running on different nodes. However, if the attacker gains access to API server credentials, they can potentially move laterally and try to compromise workloads on different nodes.
```

So without hard multi-tenant isolation between a control plane and user workloads, can a control plane truly be considered “trusted”? In the event that a workload is compromised, you can't be certain that an attacker cannot move laterally into the trusted control plane process.

## Where Gulfstream comes in

As mentioned in the introduction, Gulfstream is a streaming controller architecture for Kubernetes. It was developed to combat the cost of control plane extensions at scale, make costs proportional to usage, and amortize the resource usage over the clusters being extended.

Gulfstream encapsulates the Controller and Webhook responsibilities into CRDs. It also has a *Meta Control Plane* (MCP) that understands how to handle these CRDs and delegate the Reconciler business logic to SaaS.

> Gulfstream has two main modes of execution. Firstly, it supports users running their own MCP, also known as the "Chainguard Enforce Agent," which is useful for private clusters, or clusters behind a firewall. Chainguard also offers a fully SaaS Gulfstream, sometimes referred to as "agentless." In this case, the MCP runs on a cluster managed by Chainguard, but is pointed at a customer's cluster to work much like “agentful”. This enables a hard multi-tenant separation between our trusted control plane and the workloads running on your cluster.

Reviewer: What do you think about using "your" instead of "customer"?

```
but is pointed at one of your clusters to work much like “agentful”.
```

> Chainguard's control plane doesn’t run alongside user code, so it can provide certain security assurances that are much harder to achieve when running on the same cluster. Chainguard Enforce speaks directly to your Kubernetes API server using your cloud provider’s IAM to monitor your cluster’s workloads, and your cluster’s API server talks directly to Chainguard Enforce for admission control decisions based on policy. So even if a workload on your cluster were compromised, it is no easier for the attacker to compromise a Chainguard trusted control plane than attacking us head on.

Reviewer: ```suggestion
Chainguard's control plane doesn’t run alongside user code, so it can provide certain security assurances that are much harder to achieve when running on the same cluster. Chainguard Enforce speaks directly to your Kubernetes API server using your cloud provider’s IAM to monitor your cluster’s workloads, and your cluster’s API server talks directly to Chainguard Enforce for admission control decisions based on policy. So even if a workload on your cluster were to be compromised, it is no easier for the attacker to compromise a Chainguard trusted control plane than attacking us head on.
```

## Learn More

> As mentioned previously, running Chaingaurd Enforce in its agentless mode allows for hard multi-tenant separation between Chaingaurd's trusted control plane and your workloads. You can get started with agentless connections by first [setting up a cloud account association](../cloud-account-associations/) and then you can [use Discovery to discover your cloud resources](../chainguard-enforce-discovery-onboarding/) and then install Chainguard Enforce in Agentless mode.

Reviewer: ```suggestion
As mentioned previously, running Chainguard Enforce in its agentless mode allows for hard multi-tenant separation between Chainguard's trusted control plane and your workloads. You can get started with agentless connections by first [setting up a cloud account association](../cloud-account-associations/), then you can [use Discovery to discover your cloud resources](../chainguard-enforce-discovery-onboarding/), and finally install Chainguard Enforce in Agentless mode.
```

Author: of course the last paragraph I wrote is riddled with errors
